src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1117,6 +1140,15 @@ query_path(const char *szInput, int *is_target_path, struct dcache_rec **parent,
 			uint64_t status_old = DAOS_INIT_NOT_RUNNING;
 			bool     rc_cmp_swap;
 
+			/* Check whether MPI_Init() is running. If yes, pass to the original
+			 * libc functions. Avoid possible zeInit reentrancy/nested call.
+			 */
+
+			if (atomic_load_relaxed(&mpi_init_count) > 0) {
```

Comment:
https://en.cppreference.com/w/cpp/atomic/memory_order
From my understanding, atomic operation is already guaranteed with "memory_order_relaxed".

Reply:
Indeed you are right,  I had misunderstood the following sentence from the documentation https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync
`__ATOMIC_RELAXED
Implies no inter-thread ordering constraints.`
At the end the following documentation is really more clear on the different memory model used for inter thread synchronization https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync

However, I still have a concern with this synchronization design pattern:
```
Thread A:
Execute line 1147 and the condition is false
Stopped by the scheduler

Thread B:
Execute line 1037
Start Execute line 1038 and is interrupted by the scheduler during the execution of next_mpi_init()

Thread B:
Execute line 1158 and following
```

From my understanding we still have the race issue.

Reply:
@knard-intel Thank you very much for your comments!
The issue we encountered are in MPI applications on Aurora. The hang was caused by dead lock in nested calls of zeInit() in Intel level zero drivers in the same thread.
Both MPI_Init() and daos_init() are very expensive. We may not want to strictly exclude running them in different threads concurrently for now unless we have known issues. We do have mutex to avoid running daos_init() simultaneously in multiple threads.

Reply:
Thanks for the explanation which makes sense :-)